src/node/stage/comm.rs

```diff
@@ -0,0 +1,170 @@
+const CONNECTIONS_CACHE_SIZE: usize = 1024;
+
+// Communication component of the node to interact with other nodes.
+#[derive(Clone)]
+pub(crate) struct Comm {
+    quic_p2p: Arc<Box<QuicP2p>>,
+    endpoint: Arc<Box<Endpoint>>,
+    node_conns: Arc<Mutex<LruCache<SocketAddr, Connection>>>,
+}
+
+impl Comm {
+    pub async fn new(transport_config: Config) -> Result<Self> {
+        let quic_p2p = Arc::new(Box::new(QuicP2p::with_config(
+            Some(transport_config),
+            Default::default(),
+            true,
+        )?));
+
+        // Don't bootstrap, just create an endpoint where to listen to
+        // the incoming messages from other nodes.
+        let endpoint = Arc::new(Box::new(quic_p2p.new_endpoint()?));
+
+        let node_conns = Arc::new(Mutex::new(LruCache::with_capacity(CONNECTIONS_CACHE_SIZE)));
+
+        Ok(Self {
+            quic_p2p,
+            endpoint,
+            node_conns,
+        })
+    }
+
+    pub async fn from_bootstrapping(transport_config: Config) -> Result<(Self, Connection)> {
+        let mut quic_p2p = QuicP2p::with_config(Some(transport_config), Default::default(), true)?;
+
+        // Bootstrap to the network returning the connection to a node.
+        let (endpoint, connection) = quic_p2p.bootstrap().await?;
+
+        let quic_p2p = Arc::new(Box::new(quic_p2p));
+        let endpoint = Arc::new(Box::new(endpoint));
+        let node_conns = Arc::new(Mutex::new(LruCache::with_capacity(CONNECTIONS_CACHE_SIZE)));
+
+        Ok((
+            Self {
+                quic_p2p,
+                endpoint,
+                node_conns,
+            },
+            connection,
+        ))
+    }
+
+    /// Starts listening for connections returning a stream where to read them from.
+    pub fn listen(&mut self) -> Result<IncomingConnections> {
+        self.endpoint.listen().map_err(|err| {
+            Error::Unexpected(format!(
+                "Failed to start listening for connections: {}",
+                err
+            ))
+        })
+    }
+
+    pub fn our_connection_info(&self) -> Result<SocketAddr> {
+        self.endpoint.our_endpoint().map_err(|err| {
+            debug!("Failed to retrieve our connection info: {:?}", err);
+            err.into()
+        })
+    }
+
+    pub async fn send_message_to_targets(
+        &mut self,
+        conn_infos: &[SocketAddr],
+        delivery_group_size: usize,
+        msg: Bytes,
+    ) -> Result<()> {
+        if conn_infos.len() < delivery_group_size {
+            warn!(
+                "Less than delivery_group_size valid targets! delivery_group_size = {}; targets = {:?}; msg = {:10}",
+                delivery_group_size,
+                conn_infos,
+                HexFmt(&msg)
+            );
+        }
+
+        trace!(
+            "Sending message to {:?}",
+            &conn_infos[..delivery_group_size.min(conn_infos.len())]
+        );
+
+        // initially only send to delivery_group_size targets
```

Comment:
We should also implement the resending to the other targets when the send fails. And report failure when the resend fails too many times. And then use that to detect nodes that went offline.